### PR TITLE
Test and avoid replicate unnecessary polling over I/O futures

### PR DESCRIPTION
**Pull request by jonhoo**
See #2222.

This adds a test that replicates the issue (the I/O one; the channel one does not exhibit the issue). It then also introduces a fix by removing the (in theory) extraneous call to `wake()` in `io::Driver::register`.

### CODE REVIEW

File: tokio/src/io/driver/scheduled_io.rs [hunk 1]
```diff
@@ -56,6 +56,7 @@ impl Default for ScheduledIo {
 }
 
 impl ScheduledIo {
+    #[allow(dead_code)]
```
Comment (carllerche) [inline tokio/src/io/driver/scheduled_io.rs]: why is this needed now?

Comment (jonhoo) [reply tokio/src/io/driver/scheduled_io.rs]: Because `get_readiness` is now only called from test code, so it would issue a warning. Could make it `#[cfg(test)]`, but then various _other_ methods become unused since they were only called from `get_readiness`. So instead of going through and marking a bunch of methods as `#[cfg(test)]`, I just made this `#[allow(dead_code)]`.

**Comment by jonhoo** [comment]
@kleimkuhler Could you repeat the results of you investigation into this from Discord here?

**Comment by kleimkuhler** [comment]
@jonhoo So I don’t any performance impact results to repeat, but I did run the full Tokio and Noria test suites under this change and everything continued to pass.

File: tokio/src/io/driver/scheduled_io.rs (hunk 1 above)
Comment (carllerche) [reply tokio/src/io/driver/scheduled_io.rs]: I would prefer `#cfg(test)` we could move all of those fns to a separate impl block and annotate that block as cfg(test). Allowing dead_code will probably in that code eventually being unused and never deleted.

Comment (jonhoo) [reply tokio/src/io/driver/scheduled_io.rs]: Done

File: tokio/src/io/driver/scheduled_io.rs [hunk 2]
```diff
@@ -56,6 +56,7 @@ impl Default for ScheduledIo {
 }
 
 impl ScheduledIo {
+    #[cfg(all(test, loom))]
```
Comment (hawkw) [inline tokio/src/io/driver/scheduled_io.rs]: Hmm, if this is only used in tests now, I wonder if there's anything we can simplify based on the assumption that all readiness ops are RMWs? May be worth looking into in a follow-up...

**Comment by hawkw** [comment]
This looks good to me. It would be nice to know the history of the `wake` call to be sure it wasn't added to fix a issue that we may now see again...but since there's no bugfix commit associated with it and it was has been in `tokio` since the first commit, I doubt it was added to fix an observed bug?

**Comment by jonhoo** [comment]
@hawkw Yeah, my _guess_ (https://github.com/tokio-rs/tokio/issues/2222#issuecomment-582708850) is that it predated the contract that you _must_ continue to call poll until you get a `Pending` response. Then this wake would be necessary. I agree with you it'd be good to have justification, but not sure how we'd get it... Maybe Alex knows?

**Comment by hawkw** [comment]
@jonhoo yeah, that certainly seems like a reasonable explanation to me. Definitely not a blocker!
